plutus-use-cases/src/Plutus/Contracts/GameBabbage.hs

```diff
@@ -0,0 +1,252 @@
+
+-- | The "lock" contract endpoint. See note [Contract endpoints]
+lock :: AsContractError e => Promise () GameSchema e ()
+lock = endpoint @"lock" $ \LockArgs { lockArgsGameParam, lockArgsGameAddress, lockArgsSecret, lockArgsValue } -> do
+    logInfo @Haskell.String $ "Pay " <> Haskell.show lockArgsValue <> " to the script"
+    let lookups = Constraints.typedValidatorLookups (gameInstance lockArgsGameParam)
+        gameHash = validatorHash $ gameInstance lockArgsGameParam
+    gameRef <- findScriptReferenceByHash gameHash lockArgsGameAddress
+    let tx =  mustReferenceOutput gameRef Haskell.<>
+              Constraints.mustPayToTheScriptWithDatumInTx (hashString lockArgsSecret) lockArgsValue
+    mkTxConstraints lookups tx >>= adjustUnbalancedTx >>= yieldUnbalancedTx
+
+-- | The "guess" contract endpoint. See note [Contract endpoints]
+guess :: AsContractError e => Promise () GameSchema e ()
+guess = endpoint @"guess" $ \GuessArgs { guessArgsGameParam, guessArgsGameAddress, guessArgsSecret } -> do
+    -- Wait for script to have a UTxO of a least 1 lovelace
+    logInfo @Haskell.String "Waiting for script to have a UTxO of at least 1 lovelace"
+    utxos <- utxosAt $ gameAddress guessArgsGameParam
+    let game = gameInstance guessArgsGameParam
+        gameHash = validatorHash game
+    gameUtxos <- utxosAt guessArgsGameAddress
+    gameRef <- findScriptReferenceByHash gameHash guessArgsGameAddress
+    collateral <- fmap (head . Map.toList) ownUtxos
+    let lookups = Constraints.typedValidatorLookups (gameInstance guessArgsGameParam)
+               Haskell.<> Constraints.unspentOutputs utxos
+               Haskell.<> Constraints.unspentOutputs gameUtxos
+               Haskell.<> Constraints.unspentOutputs (uncurry Map.singleton collateral)
+        redeemer = clearString guessArgsSecret
+        tx       = Constraints.spendUtxosFromTheReferencedScript utxos redeemer gameRef
+                <> Constraints.mustUseOutputAsCollateral (fst collateral)
+    unbalancedTx <- mkTxConstraints lookups tx
+    yieldUnbalancedTx unbalancedTx
+
+-- | Get the unspent transaction outputs at an address.
+findScriptReferenceByHash ::
+    forall w s e.
+    ( AsContractError e
+    )
+    => ValidatorHash
+    -> CardanoAddress
+    -> Contract w s e TxOutRef
+findScriptReferenceByHash hash address = do
+    utxos <- utxosAt address
+    maybe
+      (throwError $ review _ContractError $ OtherContractError "Game contract not available")
+      Haskell.pure
+      $ searchReferenceScript hash utxos
+    where
+        searchReferenceScript :: ValidatorHash -> Map TxOutRef DecoratedTxOut -> Maybe TxOutRef
+        searchReferenceScript (ValidatorHash h) = let
+            getReferenceScriptHash = _2 . decoratedTxOutReferenceScript
+                . _Just . to (getScriptHash . scriptHash)
+                . only h
+            in fmap fst
+            . find (isJust . preview getReferenceScriptHash)
+            . Map.toList
```

Comment:
Maybe that should go in `Plutus.Contract.Request`?

Reply:
Actually, I was mentoning it with @james-iohk, as it can probably be reused more widely, I'd like to do it in a separate PR to add tests to it.